import os

if config == {}:
    configfile: "svtyper.json"

DATA_DIR = config["data_dir"]
SVTYPER = config["svtyper"]
AWSOUTNAME = config["awsoutname"]

SAMPLES = []
for file in os.listdir(DATA_DIR):
    if file.endswith("final.bam"):
        SAMPLES.append(str(file).split('.final.bam')[0])

print(SAMPLES)

rule all:
    input: expand("%s/{sample}.final.bam" % DATA_DIR, sample = SAMPLES), expand("{sample}.vh.vcf", sample = SAMPLES), expand("genotyped_vcfs/{sample}.vh.geno.vcf", sample = SAMPLES), expand("{sample}.gs.vcf", sample = SAMPLES), expand("genotyped_vcfs/{sample}.gs.geno.vcf", sample = SAMPLES), expand("{sample}.lumpy.vcf", sample = SAMPLES), expand("genotyped_vcfs/{sample}.lumpy.geno.vcf", sample = SAMPLES), expand("{sample}.wham.vcf", sample = SAMPLES), expand("genotyped_vcfs/{sample}.wham.geno.vcf", sample = SAMPLES), expand("{sample}.whamg.vcf", sample = SAMPLES), expand("genotyped_vcfs/{sample}.whamg.geno.vcf", sample = SAMPLES)
    params: sge_opts="-l mfree=20G -N run_all"

rule genotype_vh:
    input: "%s/{sample}.final.bam" % DATA_DIR, "{sample}.vh.vcf"
    output: "genotyped_vcfs/{sample}.vh.geno.vcf"
    log: "{sample}.vh.geno.log"
    params: sge_opts="-l mfree=10G -N geno_vh -cwd"
    shell: """
    START=$(date +%s)

    {SVTYPER} -B {input[0]} -i {input[1]} -o {output}

    END=$(date +%s)
    DIFF=$(( $END - $START ))
    echo "$DIFF" &> {log}

    """

rule genotype_gs:
    input: "%s/{sample}.final.bam" % DATA_DIR, "{sample}.gs.vcf"
    output: "genotyped_vcfs/{sample}.gs.geno.vcf"
    log: "{sample}.gs.geno.log"
    params: sge_opts="-l mfree=10G -N geno_gs -cwd"
    shell: """
    START=$(date +%s)

    {SVTYPER} -B {input[0]} -i {input[1]} -o {output}

    END=$(date +%s)
    DIFF=$(( $END - $START ))
    echo "$DIFF" &> {log}

    """

rule genotype_lumpy:
    input: "%s/{sample}.final.bam" % DATA_DIR, "{sample}.lumpy.vcf"
    output: "genotyped_vcfs/{sample}.lumpy.geno.vcf"
    log: "{sample}.lumpy.geno.log"
    params: sge_opts="-l mfree=10G -N geno_lumpy -cwd"
    shell: """
    START=$(date +%s)

    {SVTYPER} -B {input[0]} -i {input[1]} -o {output}

    END=$(date +%s)
    DIFF=$(( $END - $START ))
    echo "$DIFF" &> {log}

    """

rule genotype_whamg:
    input: "%s/{sample}.final.bam" % DATA_DIR, "{sample}.whamg.vcf"
    output: "genotyped_vcfs/{sample}.whamg.geno.vcf"
    log: "{sample}.whamg.geno.log"
    params: sge_opts="-l mfree=10G -N geno_whamg -cwd"
    shell: """
    START=$(date +%s)

    {SVTYPER} -B {input[0]} -i {input[1]} -o {output}

    END=$(date +%s)
    DIFF=$(( $END - $START ))
    echo "$DIFF" &> {log}

    """

rule genotype_wham:
    input: "%s/{sample}.final.bam" % DATA_DIR, "{sample}.wham.vcf"
    output: "genotyped_vcfs/{sample}.wham.geno.vcf"
    log: "{sample}.wham.geno.log"
    params: sge_opts="-l mfree=10G -N geno_wham -cwd"
    shell: """
    START=$(date +%s)

    {SVTYPER} -B {input[0]} -i {input[1]} -o {output}

    END=$(date +%s)
    DIFF=$(( $END - $START ))
    echo "$DIFF" &> {log}

    """


